sdk/identity/azure-identity/src/environment_credential.cpp

```diff
@@ -80,9 +164,68 @@ AccessToken EnvironmentCredential::GetToken(
 {
   if (!m_credentialImpl)
   {
-    throw AuthenticationException("EnvironmentCredential authentication unavailable. "
-                                  "Environment variables are not fully configured.");
+    std::string const AuthUnavailable = LogMsgPrefix + " authentication unavailable. ";
+
+    {
+      auto const logLevel = Logger::Level::Warning;
+      if (Log::ShouldWrite(logLevel))
+      {
+        Log::Write(
+            logLevel,
+            AuthUnavailable + "See earlier EnvironmentCredential log messages for details.");
+      }
+    }
+
+    throw AuthenticationException(
+        AuthUnavailable + "Environment variables are not fully configured.");
   }
 
   return m_credentialImpl->GetToken(tokenRequestContext, context);
 }
+
+namespace {
+void PrintCredentialCreationLogMessage(
+    std::vector<std::pair<char const*, char const*>> const& envVarsToParams,
+    char const* credThatGetsCreated)
+{
+  if (!Log::ShouldWrite(Logger::Level::Verbose))
+  {
+    if (Log::ShouldWrite(Logger::Level::Informational))
+    {
+      Log::Write(
+          Logger::Level::Informational,
+          LogMsgPrefix + " gets created with " + credThatGetsCreated + '.');
+    }
+
+    return;
+  }
+
+  auto const envVarsToParamsSize = envVarsToParams.size();
+
+  // LCOV_EXCL_START
+  AZURE_ASSERT(envVarsToParamsSize > 1);
+  // LCOV_EXCL_STOP
+
+  std::string const Tick = "'";
```

Comment:
Why use std;:string here instead of const char *?

Reply:
You're right, I'll change all but one of them to constexpr, while one of them needs to be `std::stirng` so that concatenation via `operator+` can be written.